packages/dev/core/src/Materials/shadowDepthWrapper.ts

```diff
@@ -117,39 +116,34 @@ export class ShadowDepthWrapper {
                             const subMesh = key.value;
                             if (subMesh?.getMesh() === (mesh as AbstractMesh)) {
                                 this._subMeshToEffect.delete(subMesh);
-                                const depthWrapperEntries = this._subMeshToDepthWrapper.mm.get(subMesh);
-                                if (depthWrapperEntries) {
-                                    depthWrapperEntries.forEach((depthWrapper) => {
-                                        const effect = depthWrapper.mainDrawWrapper.effect;
-                                        if (effect) {
-                                            engine._releaseEffect(effect);
-                                            effect.dispose();
-                                        }
-                                    });
-                                    this._subMeshToDepthWrapper.mm.delete(subMesh);
-                                }
+                                this._deleteDepthWrapperEffect(subMesh);
                             }
                         }
                     })
                 );
             }
 
             this._subMeshToEffect.set(params.subMesh, [params.effect, this._scene.getEngine().currentRenderPassId]);
-            const depthWrapperEntries = this._subMeshToDepthWrapper.mm.get(params.subMesh);
-            if (depthWrapperEntries) {
-                // find and release the previous depth effect
-                depthWrapperEntries.forEach((depthWrapper) => {
-                    const effect = depthWrapper.mainDrawWrapper.effect;
-                    if (effect) {
-                        engine._releaseEffect(effect);
-                        effect.dispose();
-                    }
-                });
-                this._subMeshToDepthWrapper.mm.delete(params.subMesh); // trigger a depth effect recreation
-            }
+            this._deleteDepthWrapperEffect(params.subMesh);
         });
     }
 
+    private _deleteDepthWrapperEffect(subMesh: Nullable<SubMesh>) {
+        const engine = this._scene.getEngine();
+        const depthWrapperEntries = this._subMeshToDepthWrapper.mm.get(subMesh);
+        if (depthWrapperEntries) {
+            // find and release the previous depth effect
+            depthWrapperEntries.forEach((depthWrapper) => {
+                const effect = depthWrapper.mainDrawWrapper.effect;
+                if (effect) {
+                    engine._releaseEffect(effect);

```

Comment:
> You should remove this call, as `effect.dispose()` already does it.

i tested this in my project before, it won't

Reply:
> You should remove this call, as `effect.dispose()` already does it.

it does, but there is still a big shader code string in memory, i don't kown why, so i try to call dispose() explicitly, then it works

Reply:
Are you sure the problem is coming from this? I don't see why calling first `engine._releaseEffect` would change anything... Are you able to provide a repro of the problem?

Reply:
> Are you sure the problem is coming from this? I don't see why calling first `engine._releaseEffect` would change anything... Are you able to provide a repro of the problem?

ok, i will try to build a playground tomorrow

Reply:
https://playground.babylonjs.com/#C2PIGL

i just change albedoTexure per 500ms to trigger material defines change, see `memoryLeakMaterial` and `memoryLeakTexture`, my company's project is more complex

Reply:
I tested your PR by commenting the `engine._releaseEffect(effect);` line, and `Object.keys(engine._compiledEffects).length` remains constant (=9) during the execution of the PG (?)

Also, you should investigate why we get 12 times this warning in the console log:
```markup
[.WebGL-0000054400A15B00] GL_INVALID_VALUE: Program object expected.
```
We don't have them without the PR.

Reply:
indeed, no need `engine._releaseEffect(effect);`, but i have no idea about this warning

Reply:
The fix for the warnings is a little involved (that's because the old effects should not be removed right away because of parallel shader compilation). I'm going to close this PR and will reopen a new one with your code and the fix for the warnings.

Reply:
In fact the problem was simpler, see line 126 in the other PR (#14449).